docs/source/extensions_tutorial/3_spec_api.rst

```diff
@@ -0,0 +1,173 @@
+
+Row-based tables can be specified using :py:class:`~pynwb.spec.NWBDtypeSpec`. To specify a table, provide a
+list of :py:class:`~pynwb.spec.NWBDtypeSpec` objects to the ``dtype`` argument.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBDatasetSpec, NWBDtypeSpec
+
+    spec = NWBDatasetSpec(
+        doc='A custom NWB type',
+        name='qux',
+        attribute=[
+            NWBAttributeSpec('baz', 'a value for baz', 'text'),
+            ],
+        dtype=[
+            NWBDtypeSpec('foo', 'column for foo', 'int'),
+            NWBDtypeSpec('bar', 'a column for bar', 'float')
+            ]
+        )
+
+.. tip::
+    Column-based tables are also possible and more flexible. See the documentation for `DynamicTable <https://hdmf.readthedocs.io/en/stable/tutorials/dynamictable.html>`_.
+
+Attribute Specifications
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Attributes are small metadata objects describing the nature and/or intended usage of a Group or Dataset. Attributes are
+defined in the ``attributes`` field of of a :py:class:`~pynwb.spec.NWBGroupSpec` or
+:py:class:`~pynwb.spec.NWBDatasetSpec`. ``attributes`` takes a list of :py:class:`~pynwb.spec.NWBAttributeSpec` objects.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBAttributeSpec
+
+    spec = NWBAttributeSpec(
+        name='bar',
+        doc='a value for bar',
+        dtype='float'
+    )
+
+:py:class:`~pynwb.spec.NWBAttributeSpec` has arguments very similar to :py:class:`~pynwb.spec.NWBDatasetSpec`, with
+are a few differences: ``neurodata_type_def`` and ``neurodata_type_inc`` are not allowed. An attribute cannot be a
+neurodata type. The only way to match an object with a spec is through the name of the attribute so ``name`` is
+required. You cannot have multiple attributes objects in the same place that correspond to the same
+:py:class:`~pynwb.spec.NWBAttributeSpec`, since these would have to have the same name. Therefore, instead of
+specifying number of ``quantity``, you have a ``required`` field which takes a boolean value.
+
+.. tip::
+    Dataset or Attribute? It is often possible to store data as either a Dataset or an Attribute. Our best advice is
+    to keep Attributes small, and if they are going to take any substantial amount of space, make it a Dataset.
+
+Link Specifications
+^^^^^^^^^^^^^^^^^^^
+
+You can store an object in one place and reference that object in another without copying the object using a
+`Links <https://schema-language.readthedocs.io/en/latest/specification_language_description.html#sec-link-spec>`_, which
+can be defined using :py:class:`~pynwb.spec.NWBLinkSpec` objects.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBLinkSpec
+
+    spec = NWBLinkSpec(
+        doc='my link',
+        target_type='ElectricalSeries',
+        quantity='?'
+    )
+
+``doc``, ``quantity``, and ``name`` work similarly to :py:class:`~pynwb.spec.NWBDatasetSpec`.
+
+``target_type`` indicates the neurodata type that can be referenced.
```

Comment:
```suggestion
``target_type`` indicates the neurodata type that can be referenced.

.. tip::
   In cases you need to store large collections of links, it can be more efficient to create a dataset for storing
   the links via object references. In NWB this is used, e.g,. in py:class:`~pynwb.epoch.TimeIntervals` to store 
   collections of references to TimeSeries objects. 
```